geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/utils/SortingSimpleFeatureIterator.scala

```diff
@@ -0,0 +1,289 @@
+      closed: AtomicBoolean,
+      threshold: Long): CloseableIterator[SimpleFeature] = {
+
+    // grouping ID for tmp files, doesn't have to be unique
+    lazy val id = {
+      val name = head.getFeatureType.getTypeName.replaceAll("[^A-Za-z0-9_-]", "").take(20)
+      f"$name-${Random.nextInt(10000)}%04d"
+    }
+
+    val sizable: FeatureIsSizable[SimpleFeature] = head match {
+      case _: SimpleFeature with Sizable =>
+        SizableFeatureIsSizable.asInstanceOf[FeatureIsSizable[SimpleFeature]]
+      case _ =>
+        // currently, all of our uses of this class pass in Sizable features... this means we missed one
+        logger.warn(
+          s"Feature class '${head.getClass.getName}' doesn't implement Sizable - " +
+              "using estimated size for memory threshold calculations")
+        UnSizableFeatureIsSizable
+    }
+
+    val files = ArrayBuffer.empty[File] // tmp files we create when we exceed in-memory threshold
+    lazy val serializer = KryoFeatureSerializer(head.getFeatureType, SerializationOptions.withUserData)
+    // use ArrayList for sort-in-place of the underlying array
+    val list = new java.util.ArrayList[SimpleFeature](100)
+    list.add(head)
+    var size = sizable.sizeOf(head)
+
+    // note: we already ensured tail.hasNext == true before calling this method
+    do {
+      if (size >= threshold) {
+        // write out the sorted list to disk
+        list.sort(ordering)
+        files += Files.createTempFile(s"gm-sort-$id-", ".kryo").toFile
+        logger.trace(s"Created temp sort file '${files.last.getAbsolutePath}'")
+        WithClose(new FileOutputStream(files.last)) { os =>
+          var i = 0
+          while (i < list.size()) {
+            val bytes = serializer.serialize(list.get(i))
+            os.write(ByteArrays.toBytes(bytes.length))
+            os.write(bytes)
+            i += 1
+          }
+        }
+        list.clear()
+        size = 0
+      }
+      val next = tail.next()
+      list.add(next)
+      size += sizable.sizeOf(next)
+    } while (tail.hasNext && !closed.get)
+
+    if (closed.get) {
+      files.foreach { file =>
+        if (!file.delete()) {
+          logger.warn(s"Unable to delete tmp file '${file.getAbsolutePath}''")
+          file.deleteOnExit()
+        }
+      }
+      CloseableIterator(Iterator.empty, tail.close())
+    } else {
+      if (!list.isEmpty) {
+        list.sort(ordering)
+      }
+      if (files.isEmpty) {
+        CloseableIterator(list.iterator.asScala, tail.close())
+      } else {
+        new MergeSortingIterator(files, serializer, list.iterator.asScala, tail, ordering)
+      }
+    }
+  }
+
+  /**
+   * Does a merge sort on a group of locally sorted files and a left-over in-memory iterator
+   *
+   * @param files files to merge
+   * @param serializer serializer
+   * @param mem left-over in-memory features, already sorted
+   * @param closeable link to the original feature iterator, for cleaning up on close
```

Comment:
Can we just close the original iterator after it has been exhausted?  

Isn't it likely already a CloseIterator (and hence closed by exhaustion above)?

Reply:
do you mean a SelfClosingIterator? possibly... in general I don't think we use SelfClosingIterators though. I passed it back so that any exceptions on close would be throw to the client.